<launch>
	<!-- vim: set ft=xml noet : -->
	<!-- example launch script for ArduPilot based FCU's -->

	<arg name="fcu_url" default="/dev/ttyTHS1:500000" />
	<!-- <arg name="gcs_url" default="udp://:14555@192.168.0.75:14550" /> -->
    <arg name="gcs_url" default="" />
	<arg name="tgt_system" default="1" />
	<arg name="tgt_component" default="1" />
	<arg name="log_output" default="screen" />
	<arg name="fcu_protocol" default="v2.0" />
	<arg name="respawn_mavros" default="false" />
    <arg name="set_params" default="true" />
    <arg name="imu_freq" default="200"/>
    <arg name="infra_freq" default="50"/>
    <arg name="pluginlists_yaml" default="$(find agri_resources)/config/mavros/apm_pluginlists.yaml" />
	<arg name="config_yaml" default="$(find agri_resources)/config/mavros/apm_config.yaml" />
    <arg name="self_id" default="0"/>
    <arg name="rosbag" default="false"/>

	<node pkg="mavros" type="mavros_node" name="mavros" required="$(eval not respawn_mavros)" clear_params="true" output="$(arg log_output)" respawn="$(arg respawn_mavros)">
		<param name="fcu_url" value="$(arg fcu_url)" />
		<param name="gcs_url" value="$(arg gcs_url)" />
		<param name="target_system_id" value="$(arg tgt_system)" />
		<param name="target_component_id" value="$(arg tgt_component)" />
		<param name="fcu_protocol" value="$(arg fcu_protocol)" />
		<rosparam command="load" file="$(arg pluginlists_yaml)" />
		<rosparam command="load" file="$(arg config_yaml)" />
	</node>

    <arg name="serial_no"           default=""/>
    <arg name="usb_port_id"         default=""/>
    <arg name="device_type"         default=""/>
    <arg name="json_file_path"      default=""/>
    <arg name="camera"              default="camera"/>
    <arg name="tf_prefix"           default="$(arg camera)"/>
    <arg name="external_manager"    default="false"/>
    <arg name="manager"             default="realsense2_camera_manager"/>
    <arg name="output"              default="log"/>
    <arg name="respawn"              default="false"/>

    <arg name="fisheye_width"       default="640"/>
    <arg name="fisheye_height"      default="480"/>
    <arg name="enable_fisheye"      default="false"/>

    <arg name="depth_width"         default="480"/>
    <arg name="depth_height"        default="270"/>
    <arg name="enable_depth"        default="true"/>

    <arg name="infra_width"        default="480"/>
    <arg name="infra_height"       default="270"/>
    <arg name="enable_infra1"       default="true"/>
    <arg name="enable_infra2"       default="true"/>

    <arg name="color_width"         default="640"/>
    <arg name="color_height"        default="480"/>
    <arg name="enable_color"        default="true"/>

    <arg name="fisheye_fps"         default="15"/>
    <arg name="depth_fps"           default="15"/>
    <arg name="infra_fps"           default="15"/>
    <arg name="color_fps"           default="15"/>
    <arg name="gyro_fps"            default="200"/>
    <arg name="accel_fps"           default="250"/>
    <arg name="enable_gyro"         default="false"/>
    <arg name="enable_accel"        default="false"/>

    <arg name="enable_pointcloud"         default="false"/>
    <arg name="pointcloud_texture_stream" default="RS2_STREAM_COLOR"/>
    <arg name="pointcloud_texture_index"  default="0"/>
    <arg name="allow_no_texture_points"   default="true"/>
    <arg name="ordered_pc"                default="false"/>

    <arg name="enable_sync"               default="false"/>
    <arg name="align_depth"               default="false"/>

    <arg name="publish_tf"                default="true"/>
    <arg name="tf_publish_rate"           default="0"/>

    <arg name="filters"                   default=""/>
    <arg name="clip_distance"             default="-2"/>
    <arg name="linear_accel_cov"          default="0.01"/>
    <arg name="initial_reset"             default="true"/>
    <arg name="reconnect_timeout"         default="6.0"/>
    <arg name="wait_for_device_timeout"   default="-1.0"/>
    <arg name="unite_imu_method"          default="linear_interpolation"/>
    <arg name="topic_odom_in"             default="odom_in"/>
    <arg name="calib_odom_file"           default=""/>
    <arg name="publish_odom_tf"           default="true"/>

    <group ns="$(arg camera)">
        <include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
        <arg name="tf_prefix"                value="$(arg tf_prefix)"/>
        <arg name="external_manager"         value="$(arg external_manager)"/>
        <arg name="manager"                  value="$(arg manager)"/>
        <arg name="output"                   value="$(arg output)"/>
        <arg name="respawn"                  value="$(arg respawn)"/>
        <arg name="serial_no"                value="$(arg serial_no)"/>
        <arg name="usb_port_id"              value="$(arg usb_port_id)"/>
        <arg name="device_type"              value="$(arg device_type)"/>
        <arg name="json_file_path"           value="$(arg json_file_path)"/>

        <arg name="enable_pointcloud"        value="$(arg enable_pointcloud)"/>
        <arg name="pointcloud_texture_stream" value="$(arg pointcloud_texture_stream)"/>
        <arg name="pointcloud_texture_index"  value="$(arg pointcloud_texture_index)"/>
        <arg name="enable_sync"              value="$(arg enable_sync)"/>
        <arg name="align_depth"              value="$(arg align_depth)"/>

        <arg name="fisheye_width"            value="$(arg fisheye_width)"/>
        <arg name="fisheye_height"           value="$(arg fisheye_height)"/>
        <arg name="enable_fisheye"           value="$(arg enable_fisheye)"/>

        <arg name="depth_width"              value="$(arg depth_width)"/>
        <arg name="depth_height"             value="$(arg depth_height)"/>
        <arg name="enable_depth"             value="$(arg enable_depth)"/>

        <arg name="color_width"              value="$(arg color_width)"/>
        <arg name="color_height"             value="$(arg color_height)"/>
        <arg name="enable_color"             value="$(arg enable_color)"/>

        <arg name="infra_width"              value="$(arg infra_width)"/>
        <arg name="infra_height"             value="$(arg infra_height)"/>
        <arg name="enable_infra1"            value="$(arg enable_infra1)"/>
        <arg name="enable_infra2"            value="$(arg enable_infra2)"/>

        <arg name="fisheye_fps"              value="$(arg fisheye_fps)"/>
        <arg name="depth_fps"                value="$(arg depth_fps)"/>
        <arg name="infra_fps"                value="$(arg infra_fps)"/>
        <arg name="color_fps"                value="$(arg color_fps)"/>
        <arg name="gyro_fps"                 value="$(arg gyro_fps)"/>
        <arg name="accel_fps"                value="$(arg accel_fps)"/>
        <arg name="enable_gyro"              value="$(arg enable_gyro)"/>
        <arg name="enable_accel"             value="$(arg enable_accel)"/>

        <arg name="publish_tf"               value="$(arg publish_tf)"/>
        <arg name="tf_publish_rate"          value="$(arg tf_publish_rate)"/>

        <arg name="filters"                  value="$(arg filters)"/>
        <arg name="clip_distance"            value="$(arg clip_distance)"/>
        <arg name="linear_accel_cov"         value="$(arg linear_accel_cov)"/>
        <arg name="initial_reset"            value="$(arg initial_reset)"/>
        <arg name="reconnect_timeout"        value="$(arg reconnect_timeout)"/>
        <arg name="wait_for_device_timeout"  value="$(arg wait_for_device_timeout)"/>
        <arg name="unite_imu_method"         value="$(arg unite_imu_method)"/>
        <arg name="topic_odom_in"            value="$(arg topic_odom_in)"/>
        <arg name="calib_odom_file"          value="$(arg calib_odom_file)"/>
        <arg name="publish_odom_tf"          value="$(arg publish_odom_tf)"/>

        <arg name="allow_no_texture_points"  value="$(arg allow_no_texture_points)"/>
        </include>
    </group>

    <rosparam> /camera/stereo_module/emitter_enabled: 2</rosparam>
    <!-- <rosparam> /camera/stereo_module/emitter_enabled: false</rosparam> -->
    <rosparam>
        /camera/motion_module/global_time_enabled: true
        /camera/l500_depth_sensor/global_time_enabled: true
        /camera/rgb_camera/global_time_enabled: true
    </rosparam>

    <node pkg="hardware_utils" type="hard_mavros_bridge" name="mavros_bridge" output="screen">
        <remap from="/mavros_bridge/odom" to="/d2vins/odometry"/>
        <param name="cells_batt" value="2" />
        <param name="imu_freq" value="$(arg imu_freq)" />
	    <param name="infra_freq" value="$(arg infra_freq)" />
        <param name="set_params" value="$(arg set_params)" />
        <param name="drone_id" value="$(arg self_id)" type="int"/>
        <param name="vio_align_interval" value="1000000.0" />
        <param name="debug_mode" value="false" />
        <param name="ap_param_type" value="bee" />
    </node>

    <!-- <node pkg="rosbag" type="record" name="rosbag_flight_record"
        args="record -o /tmp/flight_record 
            /mavros/imu/data_raw
            /camera/depth/color/points
            /camera/depth/image_rect_raw
            /camera/infra1/image_rect_raw
            /camera/infra2/image_rect_raw"
        if="$(arg rosbag)"/>  -->

    <node pkg="rosbag" type="record" name="rosbag_nav_record"
        args="record -o /tmp/nav_record 
            /mavros/local_position/odom
            /mavros/vision_pose/pose" 
        if="$(arg rosbag)"/> 
</launch>
